### PR TITLE
Fix `Dataframe.__setitem__` slow-downs

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
Fixes: #17140 

This PR fixes slow-downs in `DataFrame.__seitem__` by properly passing in CPU objects where needed instead of passing a GPU object and then failing and performing a GPU -> CPU transfer.

`DataFrame.__setitem__` first argument can be a column(pd.Index), in our fast path this will be converted to `cudf.Index` and thus there will be failure from cudf side and then the transfer to CPU + slow-path executes, this is the primary reason for slowdown. This PR maintains a dict mapping of such special functions where we shouldn't be converting the objects to fast path.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (Matt711) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: Why `_MethodProxy`? Is `pd.Index` pass that check?

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: That's for `__setitem__`?

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (Matt711) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1036]: What's in `custom_args[1]`?

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: `DataFrame.__seitem__` and other functions are wrapped into `_MethodProxy`

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1036]: `custom_args[0]`  has the dataframe itself and `custom_args[1]` has `arg` and `[2]` has `value`

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1036]: Thanks, the `_transform_arg` changes make sense to me

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (bdice) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:882]: Typically dunder (double-underscore) names are reserved for the Python object model. Can we get away with a single underscore, like `_customqualname`? (Would `_proxyqualname` be a better name?)

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py
Comment (Matt711) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1808]: This test would have failed before this change?

Comment (galipremsagar) [reply python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1808]: Yup

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (bdice) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1029]: I don't like `cust` as a prefix to `cust_args` here. "Custom" is too general of a word and I think we can choose a more precise/meaningful name. I am proposing `original_args` because it's pre-transformation. Other options: `input_args` or `untransformed_args`, or suggest your own.
```suggestion
                method_proxy, original_args, original_kwargs = arg
```

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (bdice) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1038]: Return this directly.
```suggestion
                return (
```

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (bdice) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1043]:
```suggestion
```

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py
Comment (bdice) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1802]: Please give another sentence or two about how this works. Specifically, that we expect the input transformation to be skipped.
```suggestion
    # We are explicitly testing the slowdown of the setitem operation.
```

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py
Comment (bdice) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1800]: Can we test this with `time.time()` instead of adding a dependency?

Is there a way to test this that doesn't require timing at all? Perhaps by mocking/augmenting the transform function to see how many times it is called?

Comment (galipremsagar) [reply python/cudf/cudf_pandas_tests/test_cudf_pandas.py]: Done.

**Comment by galipremsagar** [comment]
@bdice Addressed all your reviews.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (Matt711) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1047]:
```suggestion
```

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (Matt711) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: We only want to follow this branch when we call `DataFrame.__setitem__` and the underlying wrapped object is a `pd.DataFrame` (so we avoid the DtoH transfer), right? Would a proxy object that wraps `cudf.DataFrame` also follow this code path?

**Review by Matt711 — COMMENTED** [review]
Sorry for all the comments, just trying to make I understand the implications of this change.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: Yes and yes. This is currently written in a way we could easily add to the dict map above for any other function and parameter.

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: Can you lmk if I'm misunderstanding? If a wrapped cudf.DataFrame goes through this code path, then it will incorrectly call "_fsproxy_slow" when it should have stayed "fast". Eg. where df is a wrapped `cudf.DataFrame`
`df.__setitem__("A", pd.Index([4,5,6]))`

Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1023]: `df.__setitem__("A", pd.Index([4,5,6]))` will not force `df` into slow path. It will only for `"A"`(at 0th index) in slow path which is also `"A"`, now lets see this example:
`df.__setitem__(pd.Index(["a", "b"]), pd.Index([4,5,6]))`, this will force `pd.Index(["a", "b"])` to slow path with is a true pandas index rather than passing a cudf Index.

Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: I documented a bit more thoroughly here: https://github.com/rapidsai/cudf/pull/17222/commits/04ec5b7139eac2db8e7b87b03a5fec104a285808

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: Thanks for adding that! I have a question about this example:
```
df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))
```
So `df` can either be a wrapped a pandas or cudf dataframe.
1. `df` is a wrapped a pandas dataframe. 
In this case, `pd.Index(["a", "b"])` is forced to take the slow path. This makes sense to because if we keep it a fast-slow proxy object, then it will be converted to `cudf.Index(["a", "b"])` and then cause `df.__setitem__(cudf.Index(["a", "b"]), ...)` to fail because you can't pass a GPU object to pandas like this. Because this would fail, we'd trigger a DtoH transfer. This data transfer is the main contributing factor to the slow down. This case makes sense to me.
2.`df` is a wrapped a cudf dataframe. 
In this case, `pd.Index(["a", "b"])` is also forced to take the slow path. Is this okay because `__setitem__` also fails in this case? Trying this locally, I get
```
TypeError: Index object is not iterable. Consider using `.to_arrow()`, `.to_pandas()` or `.values_host` if you wish to iterate over the values.
```

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (Matt711) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:45]:
```suggestion
# This is a dict of functions that are known to have arguments that
# need to be transformed from fast to slow only. i.e., Some cudf functions
# error on passing a device object but don't error on passing a host object.
# For example: DataFrame.__setitem__(arg, value) errors on passing a
# cudf.Index object but doesn't error on passing a pd.Index object.
# Hence we need to transform the arg from fast to slow only. So, we use
# a dictionary like:
# {"DataFrame.__setitem__": {0}}
# where the keys are the function names and the values are the indices
# (0-based) of the arguments that need to be transformed.
```

**Review by Matt711 — COMMENTED** [review]
This is making more sense to me.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: Were you arriving at error 2 via public api or mocking the api? I think error 2 isn't possible via public apis.

Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: For the following case:

df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

It will be converted to this and tried with cudf:

df.__setitem__(pd.Index(["a", "b"]), cudf.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

And then if the above fails the following will be tried on pandas:

df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: No it doesn't fail via the public API, I used `CUDF_PANDAS_FALLBACK_DEBUGGING=True`

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: It turns out (2.) doesn't work even w/o `cudf.pandas`

```
In [1]: import cudf

In [2]: import pandas as pd

In [3]: df = cudf.DataFrame()

In [4]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

ValueError: Data must be 1-dimensional

In [5]: df = pd.DataFrame()

In [6]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

In [7]: df
Out[7]: 
   a   b
0  4  10
1  5  11
2  6  12
```

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: Yep there's fallback in this case anyway.
```
In [1]: %load_ext cudf.pandas

In [2]: import pandas as pd

In [3]: df = pd.DataFrame()

In [4]: type(df._fsproxy_wrapped)
Out[4]: cudf.core.dataframe.DataFrame

In [5]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

In [6]: type(df._fsproxy_wrapped)
Out[6]: pandas.core.frame.DataFrame
```

Comment (Matt711) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:1034]: This what I get with your changes. (And running with `CUDF_PACDAS_DEBUGGING=True` does not fail). This looks correct to me
```
In [1]: %load_ext cudf.pandas

In [2]: import pandas as pd

In [3]: df = pd.DataFrame()

In [4]: type(df._fsproxy_wrapped)
Out[4]: cudf.core.dataframe.DataFrame

In [5]: df.__setitem__(pd.Index(["a", "b"]), pd.DataFrame({'a':[4,5,6], 'b':[10, 11, 12]}))

In [6]: type(df._fsproxy_wrapped)
Out[6]: cudf.core.dataframe.DataFrame
```

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py
Comment (Matt711) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1802]:
```suggestion
    # We are explicitly testing the slowdown of the setitem operation
    # by eliminating the DtoH transfer performed by df[df.columns] = ...
    # We do this by ensuring the df.columns argument in the setitem
    # operation remains a slow object.
```

**Review by Matt711 — APPROVED** [review]
Thanks @galipremsagar for explaining that. I'm not sure if you want to wait for another review, but it LGTM.

**Comment by galipremsagar** [comment]
> Thanks @galipremsagar for explaining that. I'm not sure if you want to wait for another review, but it LGTM.

Thanks for the review, Matt! I'll go ahead and merge. I'll also open a follow-up to fix the cudf bug you surfaced in this discussion.

**Pull request merged**
